File: csharp/strain/Strain.cs
using System;
using System.Collections.Generic;

public static class Strain
{
    public static IEnumerable<T> Keep<T>(this IEnumerable<T> collection, Func<T, bool> predicate)
    {
        foreach (var n in collection)
        {
            if (predicate(n))
            {
                yield return n;
            }
        }
    }

    public static IEnumerable<T> Discard<T>(this IEnumerable<T> collection, Func<T, bool> predicate)
    {
        foreach (var n in collection)
        {
            if (!predicate(n))
            {
                yield return n;
            }
        }
    }
}

File: lua/strain/strain.lua
local function keep(lst, pred)
    local out = {}
    for _, v in ipairs(lst) do if pred(v) then out[#out + 1] = v end end
    return out
end
return {keep = keep, discard = function (lst, pred) return keep(lst, function(x) return not pred(x) end) end }
